src/main/java/seedu/algobase/storage/JsonAdaptedProblemSearchRule.java

```diff
@@ -0,0 +1,137 @@
+     */
+    public JsonAdaptedProblemSearchRule(ProblemSearchRule rule) {
+        name = rule.getName().name;
+        if (!rule.hasDefaultNamePredicate()) {
+            this.nameContainsKeywordsPredicate = new JsonAdaptedNameContainsKeywordsPredicate(rule.getNamePredicate());
+        } else {
+            this.nameContainsKeywordsPredicate = null;
+        }
+        if (!rule.hasDefaultAuthorPredicate()) {
+            this.authorMatchesKeywordPredicate =
+                new JsonAdaptedAuthorMatchesKeywordPredicate(rule.getAuthorPredicate());
+        } else {
+            this.authorMatchesKeywordPredicate = null;
+        }
+        if (!rule.hasDefaultDescriptionPredicate()) {
+            this.descriptionContainsKeywordsPredicate =
+                new JsonAdaptedDescriptionContainsKeywordsPredicate(rule.getDescriptionPredicate());
+        } else {
+            this.descriptionContainsKeywordsPredicate = null;
+        }
+        if (!rule.hasDefaultSourcePredicate()) {
+            this.sourceMatchesKeywordPredicate =
+                new JsonAdaptedSourceMatchesKeywordPredicate(rule.getSourcePredicate());
+        } else {
+            this.sourceMatchesKeywordPredicate = null;
+        }
+        if (!rule.hasDefaultDifficultyPredicate()) {
+            this.difficultyIsInRangePredicate =
+                new JsonAdaptedDifficultyIsInRangePredicate(rule.getDifficultyPredicate());
+        } else {
+            this.difficultyIsInRangePredicate = null;
+        }
+        if (!rule.hasDefaultTagPredicate()) {
+            this.tagIncludesKeywordsPredicate =
+                new JsonAdaptedTagIncludesKeywordsPredicate(rule.getTagPredicate());
+        } else {
+            this.tagIncludesKeywordsPredicate = null;
+        }
+    }
+
+
+    /**
+     * Converts a name in string format to a Name Object.
+     *
+     * @param name name in string format.
+     * @return the corresponding Name Object.
+     * @throws IllegalValueException if string format is invalid.
+     */
+    public Name retrieveName(String name) throws IllegalValueException {
+        if (name == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Name.class.getSimpleName()));
+        }
+        if (!Name.isValidName(name)) {
+            throw new IllegalValueException(String.format(Name.MESSAGE_CONSTRAINTS));
+        }
+        return new Name(name);
+    }
+
+    /**
+     * Converts this Jackson-friendly adapted Problem object into the model's
+     * {@code ProblemSearchRule} object.
+     *
+     * @throws IllegalValueException if there were any data constraints violated in the adapted
+     * ProblemSearchRule.
+     */
+    public ProblemSearchRule toModelType() throws IllegalValueException {
+        final Name name = retrieveName(this.name);
+        final NameContainsKeywordsPredicate namePredicate = this.nameContainsKeywordsPredicate == null ? null
```

Comment:
I think it might be cleaner to define a method instead of using ternary operators

Reply:
They are of different types so I don't think I can write one method to handle all these... I'll change them to `if...else` for now. Do you have better ideas?

Reply:
One way is for all of them to implement a `Field` interface with a `toModelType()` method, but it  may not be worth refactoring as `if...else` should be sufficient